### DISCUSSION TITLE
Wrong id in insertOnConflictUpdate

### DISCUSSION THREAD

**Discussion by arefrafei92** [simolus3/drift#1165] [Q&A]
Hey,
I'm using moor for a long time, but I have a weird issue recently.

I have a table with primaryKey like below :  

```
@DataClassName("Book")
class Books extends Table {
  @override
  Set<Column> get primaryKey => {id};
  IntColumn get id => integer()();
  TextColumn get title => text()();
  ...
```

and in BooksDao I have this function : 
```
  void addToBookmark(Book entry, {bool isLiked = true}) {
    into(books)
        .insertOnConflictUpdate(entry.copyWith( isLiked: isLiked))
        .then((value) => logger.w("entry Id : ${entry.id} / db Id : $value"));
  }
```

as you can see we expect to see same result for 'entry Id' and 'db Id'.

first time that I add a book to bookmark it's correct and when I'm in same page it works good but when I go back and re-enter again and tap on addToBookmark I got different Ids for output.

I don't know what's problem

I access to Dao like this : 

`final BooksDao dbHelper = BooksDao(locator<MyDatabase>());
`
I used Provider as well.

It seems when I make an instance for Dao I have a problem, if the book wasn't in database so it's not conflict and ids are correct .

**Comment by kuhnroyal** [comment]
Read the docs for `insertOnConflictUpdate`, I believe it only guarantees a correct ID for an actual insert.

**Reply by arefrafei92** [reply]
I know and it should be 

but after re-run my application and use insertOnConflictUpdate I got another id which isn't related to any item 🤔

**Reply by simolus3** [reply]
When `insertOnConflictUpdate` behaves as an update (that is, there's a conflicting row), it does not set the last insert id in sqlite3. `insert` returns that value though, so you essentially get nonsense (or rather, the id of the last insert which might have happened in another table).

**Reply by arefrafei92** [reply]
Is there any way to check what's problem ? 

as I found it's updating last item no matter what id ( primary key) I'm sending at that time.
I'm using StreamBuilder to get data and when I remove an item StreamBuilder can't return last item and it seems that I change last item !
but if I close and go to page again, I can see all items.

I wish I could make a video to send you

**Reply by simolus3** [reply]
> as I found it's updating last item no matter what id ( primary key) I'm sending at that time.

Hm, that doesn't sound right. Can you enable `logStatements: true` on the `VmDatabase` and post the generated sql?
